# Carregar e organizar um conjunto de dados usando o pacote pandas


## Links para dados de seguro de automóveis da SUSEP

* Autoseg: https://www2.susep.gov.br/menuestatistica/Autoseg/principal.aspx


Sobre os dados completos:
* [Manual da SUSEP](https://www.gov.br/susep/pt-br/servicos/mercado/enviar-dados/arquivos/manual_de_orientacao_para_envio_de_dados_out2023.pdf) (p.81-. Sinistros: p.87-)


* Link para baixar os dados completos: ['AUTO_2021A.rar'](https://www2.susep.gov.br/download/basesdados/scd_auto/AUTO_2021A.rar)

Após baixar os dados completos no computador, eu descompactei o arquivo, o que gerou:
* R_AUTO_2021A: apólices em 2020
* S_AUTO_2021A: sinistros em 2020

Eu carreguei o arquivo 'S_AUTO_2021A.csv' para o meu *google drive*.

* OBS.: Tirei do meu google drive devido ao tamanho do arquivo (Se você tiver dificuldades para baixar o arquivo .rar acima, tente baixar diretamente aqui: [S_AUTO_2021A.csv](https://drive.google.com/file/d/18a9l66HZpLjKRki45X9_udM6iFfGlozy/view?usp=drive_link))



Importar pacotes necessários:

In [5]:
import pandas as pd
import numpy as np
from google.colab import drive, files
import os

pd.options.display.float_format = '{:.2f}'.format    # pandas: para todos os números aparecerem com duas casas decimais

## Leitura dos microdados

Os dados serão lidos a partir do *google drive*.

In [6]:
# alterar o caminho se necessário, especificando a pasta
drive.mount('/content/drive')
os.chdir('/content/drive/My Drive/dados/AUTO_2021A')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [7]:
# ler o arquivo csv
# o conjunto de dados está com o separador ; e os decimais com ,
df = pd.read_csv('S_AUTO_2021A.csv', sep=';', decimal=',')

<ipython-input-7-39fe32745602>:3: DtypeWarning: Columns (1,9) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('S_AUTO_2021A.csv', sep=';', decimal=',')


## Visão geral dos dados


In [8]:
# mostrar as 5 primeiras linhas
df.head()

cod_apo     ENDOSSO  ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  \
0      NaN  0000000000     0           1          1          1   004354-0   
1      NaN  0000000000     0           1          1          1   004354-0   
2      NaN  0000000000     0           1          1          1   004354-0   
3      NaN  0000000000     0           1          1          1   004354-0   
4      NaN  0000000000     0           1          1          1   004354-0   

   ANO_MODELO COD_TARIF REGIAO  ...  D_SALVADO  VAL_RESS  D_RESS     D_AVI  \
0        2009        10     11  ...          0      0.00       0  20201016   
1        2009        10     11  ...          0      0.00       0  20201017   
2        2009        10     11  ...          0      0.00       0  20201017   
3        2009        10     11  ...          0      0.00       0  20200812   
4        2009        10     11  ...          0      0.00       0  20200812   

      D_LIQ   D_OCORR  CAUSA  SEXO    D_NASC      CEP  
0  20201026  20201015      7     M  19740216  9371420  
1  20201026  20201016      7     M  19740216  9371420  
2  20201105  20201016      7     M  19740216  9030620  
3  20200819  20200811      7     M  19740216  9371520  
4  20200819  20200811      7     M  19740216        0  

[5 rows x 24 columns]

In [9]:
# número de linhas (observações) e colunas (variáveis)
df.shape

(3747647, 24)

Há informações sobre 3.747.647 observações (sinistros) e são 24 variáveis.

In [10]:
# nomes das variáveis
df.columns

Index(['cod_apo', 'ENDOSSO', 'ITEM', 'MODALIDADE', 'TIPO_PROD', 'COBERTURA',
       'COD_MODELO', 'ANO_MODELO', 'COD_TARIF', 'REGIAO', 'COD_CONT', 'EVENTO',
       'INDENIZ', 'VAL_SALVAD', 'D_SALVADO', 'VAL_RESS', 'D_RESS', 'D_AVI',
       'D_LIQ', 'D_OCORR', 'CAUSA', 'SEXO', 'D_NASC', 'CEP'],
      dtype='object')

In [11]:
# informações sobre os tipos de dados
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3747647 entries, 0 to 3747646
Data columns (total 24 columns):
 #   Column      Dtype  
---  ------      -----  
 0   cod_apo     float64
 1   ENDOSSO     object 
 2   ITEM        int64  
 3   MODALIDADE  int64  
 4   TIPO_PROD   int64  
 5   COBERTURA   int64  
 6   COD_MODELO  object 
 7   ANO_MODELO  int64  
 8   COD_TARIF   object 
 9   REGIAO      object 
 10  COD_CONT    int64  
 11  EVENTO      int64  
 12  INDENIZ     float64
 13  VAL_SALVAD  float64
 14  D_SALVADO   int64  
 15  VAL_RESS    float64
 16  D_RESS      int64  
 17  D_AVI       int64  
 18  D_LIQ       int64  
 19  D_OCORR     int64  
 20  CAUSA       int64  
 21  SEXO        object 
 22  D_NASC      int64  
 23  CEP         int64  
dtypes: float64(4), int64(15), object(5)
memory usage: 686.2+ MB


In [12]:
# para visualizar até 50 colunas ou o número desejado
pd.set_option('display.max_columns', 24)

In [13]:
# mostrar as 5 últimas linhas
df.tail()

cod_apo     ENDOSSO  ITEM  MODALIDADE  TIPO_PROD  COBERTURA  \
3747642 6537392.00  0000000000     1           1          1          1   
3747643 6537421.00  0000000000     1           1          1          1   
3747644 6537507.00  0000000000     1           1          1          1   
3747645 6537566.00  0000000000     1           1          1          1   
3747646 6537566.00  0000000000     1           1          1          1   

        COD_MODELO  ANO_MODELO COD_TARIF REGIAO  COD_CONT  EVENTO  INDENIZ  \
3747642   001414-1        2015       14A     36         1       2  2300.00   
3747643   003271-9        2011        10     04         1       2     0.00   
3747644   001203-3        2012       14A     04         1       2   470.00   
3747645   004301-0        2010        22     04         1       2  3479.00   
3747646   004301-0        2010        22     04         1       2  4440.00   

         VAL_SALVAD  D_SALVADO  VAL_RESS  D_RESS     D_AVI     D_LIQ  \
3747642        0.00          0      0.00       0  20200619  20200817   
3747643        0.00          0      0.00       0  20200609  20200626   
3747644        0.00          0      0.00       0  20201125  20210119   
3747645        0.00          0      0.00       0  20200914  20201007   
3747646        0.00          0      0.00       0  20200616  20200720   

          D_OCORR  CAUSA SEXO    D_NASC       CEP  
3747642  20200513      4    0         0  78705000  
3747643  20200529      4    M  19510321  89715000  
3747644  20201122      4    M  19730527  89915000  
3747645  20200912      4    M  19680306  89715000  
3747646  20200616      4    M  19680306  89715000

In [14]:
# mostrar uma parte aleatória das linhas
df.sample(10)

cod_apo     ENDOSSO  ITEM  MODALIDADE  TIPO_PROD  COBERTURA  \
3745573  1044281.00  0000000000     0           1          2          1   
680132          NaN  0000000000     0           1          1          1   
961026  10019377.00  0000000000     1           1          2          1   
3202075         NaN  0000000000   264           1          1          1   
162075          NaN  0000000000     0           1          1          1   
3618322  9137591.00  0000000000     1           1          2          1   
3587548  1301734.00  0000000000     0           1          2          1   
1957828 21950887.00  0000000000     1           1          2          1   
2141320   871203.00  0000000000     0           1          2          1   
1309117 12954306.00           0     0           1          1          1   

        COD_MODELO  ANO_MODELO COD_TARIF REGIAO  COD_CONT  EVENTO  INDENIZ  \
3745573   004338-9        2009       14A     20         1       1   503.00   
680132    005365-1        2016        10     11         1       8   220.00   
961026    014043-0        2005        10     13         1       1 18481.00   
3202075   005459-3        2019       14C     13         1       1   468.00   
162075    033062-0        2017        23     11         1       8   103.00   
3618322   004001-0        2000        10     12         1       1   124.00   
3587548   025244-1        2018        10     36         1       1   497.00   
1957828   003363-4        2020        22     16         1       8   567.00   
2141320   029039-4        2015        11     16         1       1   497.00   
1309117   827102-0        2018        30     39         1       8    97.00   

         VAL_SALVAD  D_SALVADO  VAL_RESS  D_RESS     D_AVI     D_LIQ  \
3745573        0.00          0      0.00       0  20201114         0   
680132         0.00          0      0.00       0  20201009  20201030   
961026      8400.00   20210122      0.00       0  20201217         0   
3202075        0.00          0      0.00       0  20201203  20210502   
162075         0.00          0      0.00       0  20200723  20200812   
3618322        0.00          0      0.00       0  20200810         0   
3587548        0.00          0      0.00       0  20201204         0   
1957828        0.00          0      0.00       0  20201215  20210107   
2141320        0.00          0      0.00       0  20201214         0   
1309117        0.00          0      0.00       0  20201022  20201105   

          D_OCORR  CAUSA SEXO    D_NASC       CEP  
3745573  20201114      2    M  19780512  29167000  
680132   20201007      7    M  19520802  11672020  
961026   20201217      9    F  19900223  13423650  
3202075  20201201      9    0         0  17512050  
162075   20200721      7    M  19770114   4619000  
3618322  20200806      9    M  19630213  13063170  
3587548  20201204      2    F  19770512  78118000  
1957828  20201208      9    F  19760408  33858600  
2141320  20201214      2    F  19811207  30580000  
1309117  20201021      7    M  19800220  74922480

Obter o resumo estatístico das variáveis auxilia no entendimento delas. É possível saber a média, mínimo, máximo e quartis (25%, 50% e 75%):

In [15]:
df.describe()

cod_apo       ITEM  MODALIDADE  TIPO_PROD  COBERTURA  ANO_MODELO  \
count  2615880.00 3747647.00  3747647.00 3747647.00 3747647.00  3747647.00   
mean  10679862.69      66.81        1.04       1.47       1.20     2011.98   
std    6156638.10    1455.97        0.27       0.50       1.15       64.34   
min          4.00       0.00        0.00       0.00       0.00        0.00   
25%    5608399.00       0.00        1.00       1.00       1.00     2012.00   
50%   10329021.50       0.00        1.00       1.00       1.00     2014.00   
75%   13754533.00       1.00        1.00       2.00       1.00     2018.00   
max   23744956.00  180494.00        4.00       2.00       9.00     2022.00   

        COD_CONT     EVENTO    INDENIZ  VAL_SALVAD   D_SALVADO   VAL_RESS  \
count 3747647.00 3747647.00 3747647.00  3747647.00  3747647.00 3747647.00   
mean        1.01       3.86    3157.45      253.76   286923.39       5.69   
std         0.09       3.35   12254.93     3677.92  2390552.54     545.28   
min         0.00       0.00 -176275.00  -414000.00        0.00  -75980.00   
25%         1.00       1.00      98.00        0.00        0.00       0.00   
50%         1.00       1.00     243.00        0.00        0.00       0.00   
75%         1.00       8.00    1225.00        0.00        0.00       0.00   
max         2.00       8.00 1457500.00   454777.00 20220301.00  215484.00   

           D_RESS       D_AVI       D_LIQ     D_OCORR      CAUSA      D_NASC  \
count  3747647.00  3747647.00  3747647.00  3747647.00 3747647.00  3747647.00   
mean     24204.87 20201240.90 15095704.60 20200955.08       6.75 16779543.75   
std     698951.46    33043.55  8780685.93    14809.42       2.27  7013887.32   
min          0.00        0.00        0.00        0.00       0.00        0.00   
25%          0.00 20200819.00        0.00 20200814.00       4.00 19530221.00   
50%          0.00 20201007.00 20201002.00 20201001.00       7.00 19681218.00   
75%          0.00 20201123.00 20201210.00 20201117.00       9.00 19810311.00   
max   20220303.00 20210228.00 20220317.00 20210227.00       9.00 29950928.00   

              CEP  
count  3747647.00  
mean  33692749.73  
std   32311580.06  
min          0.00  
25%    6140040.00  
50%   18540000.00  
75%   60822000.00  
max   99999999.00

## Alguns ajustes nos dados

Antes de começar a alterar algumas coisas no conjunto de dados, vamos realizar uma cópia deles com o nome 'dados':

In [16]:
dados = df.copy()

Verificar se há dados faltantes:

In [17]:
# sem sum() temos apenas True e False
dados.isna().sum()

cod_apo       1131767
ENDOSSO             0
ITEM                0
MODALIDADE          0
TIPO_PROD           0
COBERTURA           0
COD_MODELO          1
ANO_MODELO          0
COD_TARIF           0
REGIAO              0
COD_CONT            0
EVENTO              0
INDENIZ             0
VAL_SALVAD          0
D_SALVADO           0
VAL_RESS            0
D_RESS              0
D_AVI               0
D_LIQ               0
D_OCORR             0
CAUSA               0
SEXO                0
D_NASC              0
CEP                 0
dtype: int64

Há 1.131.767 dados faltantes de 'cod_apo'.

Num primeiro momento, 'cod_apo' e 'ENDOSSO' não serão úteis para nossas análises, vamos retirá-las do conjunto de dados.

In [18]:
# apagar variáveis
dados = dados.drop(['cod_apo', 'ENDOSSO'], axis=1)

In [19]:
# identificar a linha com dados faltantes
dados[dados.isnull().any(axis=1)]

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO  \
2398850     0           0          0          0        NaN           0   

        COD_TARIF REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  \
2398850         0     00         0       0     0.00        0.00          0   

         VAL_RESS  D_RESS  D_AVI  D_LIQ  D_OCORR  CAUSA SEXO  D_NASC  CEP  
2398850      0.00       0      0      0        0      0    0       0    0

Como essa linha não possui nenhuma informação relevante, vamos retirá-la.

In [20]:
dados.dropna(inplace=True)

In [21]:
dados.shape

(3747646, 22)

## Observar variáveis

### 'REGIAO'

As regiões do Brasil são codificadas em números de 1 a 41. Vamos ver como está a variável 'REGIAO' no conjunto de dados.

In [22]:
# visualizar todas as ocorrências para 'REGIAO'
dados['REGIAO'].unique()

array([11, 9, 18, 30, 13, 16, 21, 20, 1, 15, 14, 7, 5, 4, 37, 36, 19, 39,
       2, 23, 27, 17, 24, 8, 38, 28, 26, 3, 29, 12, 99, 22, 31, 6, 10, 40,
       34, 41, 35, 0, 25, 32, 33, '02', '13', '14', '38', '03', '01',
       '18', '19', '11', '10', '39', '04', '09', '37', '05', '08', '06',
       '16', '24', '21', '36', '15', '12', '35', '07', '41', '40', '20',
       '25', '29', '23', '99', '28', '00', '26', '27', '17', '30', '22',
       '31', '33', '34', ' .', '32', 'PI', 'MG', 'RJ', 'SC', 'PR', 'GO',
       'RS', 'ES', 'AM', 'CE', 'MS', 'PE', 'DF', 'PA', 'BA'], dtype=object)

Algumas ocorrências estão sendo lidas como strings e outras como inteiros. Por exemplo: temos 2 e '02'. Há observações com outros valores no lugar do código da região: ' .', '00', '99', '0'.

Há algumas observações com as siglas dos estados no lugar do código da região.

In [23]:
# converter todas as ocorrências para número e considerar as não númericas como NaN
dados['REGIAO'] = pd.to_numeric(dados['REGIAO'], errors='coerce')

In [24]:
# visualizar todas as ocorrências para 'REGIAO'
dados['REGIAO'].unique()

array([11.,  9., 18., 30., 13., 16., 21., 20.,  1., 15., 14.,  7.,  5.,
        4., 37., 36., 19., 39.,  2., 23., 27., 17., 24.,  8., 38., 28.,
       26.,  3., 29., 12., 99., 22., 31.,  6., 10., 40., 34., 41., 35.,
        0., 25., 32., 33., nan])

Agora vamos transformar 'nan' em 0 também para sinalizar que não há informação sobre região:

In [25]:
dados['REGIAO'].fillna(0, inplace=True)

<ipython-input-25-fa0ef8e343c1>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dados['REGIAO'].fillna(0, inplace=True)


In [26]:
# visualizar todas as ocorrências para 'REGIAO'
dados['REGIAO'].unique()

array([11.,  9., 18., 30., 13., 16., 21., 20.,  1., 15., 14.,  7.,  5.,
        4., 37., 36., 19., 39.,  2., 23., 27., 17., 24.,  8., 38., 28.,
       26.,  3., 29., 12., 99., 22., 31.,  6., 10., 40., 34., 41., 35.,
        0., 25., 32., 33.])

In [27]:
# número de ocorrências de cada valor da variável 'REGIAO'
dados['REGIAO'].value_counts()

REGIAO
11.00    995738
13.00    412541
16.00    193870
18.00    184816
21.00    148604
12.00    130399
7.00     129540
9.00     116569
2.00     114680
1.00     113977
8.00      89786
23.00     84417
39.00     83637
5.00      80704
38.00     73071
10.00     67535
3.00      63179
19.00     62457
14.00     55157
20.00     54358
15.00     53920
27.00     50280
36.00     45600
37.00     37006
4.00      33665
25.00     30956
30.00     29934
24.00     29687
6.00      27589
26.00     21460
22.00     21310
99.00     17811
17.00     17342
29.00     17214
41.00     14500
28.00     13328
40.00     10321
31.00      9756
33.00      3442
0.00       3076
35.00      2315
32.00      1053
34.00      1046
Name: count, dtype: int64

Depois podemos tentar usar a informação do CEP para recuperar de qual localidade é cada sinistro.

Vamos fazer uma consulta aos dados para ver como são os sinistros cuja região é 0.

In [28]:
# se quisermos ver como são as observações com 0, por exemplo
dados.query("REGIAO == 0")

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO  \
1712        1           1          2          1   001449-4        2020   
1713        1           1          2          1   001449-4        2020   
7268        1           1          2          1   005345-7        2012   
7269        1           1          2          1   005345-7        2012   
7270        1           1          2          1   005345-7        2012   
...       ...         ...        ...        ...        ...         ...   
3743628     1           1          2          1   003318-9        2011   
3743629     1           1          2          1   003318-9        2011   
3743630     1           1          2          1   003318-9        2011   
3743631     1           1          2          1   003318-9        2011   
3743632     1           1          2          1   003318-9        2011   

        COD_TARIF  REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  \
1712           22    0.00         1       1 16482.00        0.00          0   
1713           22    0.00         1       2   716.00        0.00          0   
7268           10    0.00         1       1 20936.00        0.00          0   
7269           10    0.00         1       2 11536.00        0.00          0   
7270           10    0.00         1       2 11536.00        0.00          0   
...           ...     ...       ...     ...      ...         ...        ...   
3743628        10    0.00         1       1  1402.00        0.00          0   
3743629        10    0.00         1       1     0.00   -20200.00   20210201   
3743630        10    0.00         1       1   180.00        0.00          0   
3743631        10    0.00         1       1    26.00        0.00          0   
3743632        10    0.00         1       2  5183.00        0.00          0   

         VAL_RESS  D_RESS     D_AVI     D_LIQ   D_OCORR  CAUSA SEXO    D_NASC  \
1712         0.00       0  20200714  20200928  20200708      4    M  19720511   
1713         0.00       0  20200812  20200922  20200708      9    M  19720511   
7268         0.00       0  20201223  20210115  20201222      5    F  19811229   
7269         0.00       0  20201223  20210201  20201222      9    F  19811229   
7270         0.00       0  20201223  20210205  20201222      9    F  19811229   
...           ...     ...       ...       ...       ...    ...  ...       ...   
3743628      0.00       0  20201124  20201222  20201121      9    F  19760216   
3743629      0.00       0  20201124  20201222  20201121      9    F  19760216   
3743630      0.00       0  20201124  20201222  20201121      9    F  19760216   
3743631      0.00       0  20201124  20201222  20201121      9    F  19760216   
3743632      0.00       0  20201124  20210108  20201121      9    F  19760216   

              CEP  
1712     44004152  
1713     44004152  
7268     38307783  
7269     38307783  
7270     38307783  
...           ...  
3743628  38412878  
3743629  38412878  
3743630  38412878  
3743631  38412878  
3743632  38412878  

[3076 rows x 22 columns]

In [29]:
# ver se há CEPs com valor 0
dados.query("CEP == 0")

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO  \
4           0           1          1          1   004354-0        2009   
11          0           1          1          1   023113-4        2013   
14          0           1          1          1   001378-1        2013   
97          0           1          2          1   001291-2        2016   
98          0           1          2          1   004424-5        2018   
...       ...         ...        ...        ...        ...         ...   
3747543     0           1          2          1   011198-8        2020   
3747544     0           1          2          1   003414-2        2015   
3747545     0           1          2          1   999999-9        2015   
3747546     0           1          2          1   999999-9        2015   
3747547     0           1          2          1   001314-5        2014   

        COD_TARIF  REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  \
4              10   11.00         1       8   127.00        0.00          0   
11             10   11.00         1       8   126.00        0.00          0   
14             10   11.00         1       8   127.00        0.00          0   
97            14A   19.00         1       1   452.00        0.00          0   
98             10   19.00         1       1  2342.00        0.00          0   
...           ...     ...       ...     ...      ...         ...        ...   
3747543        10    5.00         1       1  6197.00        0.00          0   
3747544        10   39.00         1       1  5798.00        0.00          0   
3747545        10    1.00         1       1  7954.00        0.00          0   
3747546        10    1.00         1       2  4127.00        0.00          0   
3747547        10    2.00         1       1  3490.00        0.00          0   

         VAL_RESS  D_RESS     D_AVI     D_LIQ   D_OCORR  CAUSA SEXO    D_NASC  \
4            0.00       0  20200812  20200819  20200811      7    M  19740216   
11           0.00       0  20200922  20200925  20200903      7    M  19790430   
14           0.00       0  20200818  20200909  20200817      7    M  19580328   
97           0.00       0  20200525         0  20200523      4    M  19610117   
98           0.00       0  20200908         0  20200907      4    F  19490508   
...           ...     ...       ...       ...       ...    ...  ...       ...   
3747543      0.00       0  20201229         0  20201223      4    M  19890628   
3747544      0.00       0  20210107         0  20201230      9    M  19771206   
3747545      0.00       0  20201221         0  20201218      4    F  19680714   
3747546      0.00       0  20201221         0  20201218      4    F  19680714   
3747547      0.00       0  20201124         0  20201122      4    F  19760531   

         CEP  
4          0  
11         0  
14         0  
97         0  
98         0  
...      ...  
3747543    0  
3747544    0  
3747545    0  
3747546    0  
3747547    0  

[340558 rows x 22 columns]

Há muitas observações com CEP igual a zero. Será que há observações com CEP igual a 0 e região com valor 0 também? Ou seja, há observações sem informação alguma sobre a localidade?

In [30]:
# visualizar linhas com REGIAO igual a 0 e CEP igual a 0
dados.query('REGIAO == 0 & CEP == 0')

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO  \
29749       2           1          1          1   999999-9        2007   
43148       1           1          1          1   999999-9           0   
44185       1           1          1          1   999999-9           0   
69066       1           3          1          9   999999-9        1997   
96831      13           2          1          1   999999-9        1975   
...       ...         ...        ...        ...        ...         ...   
3698859    27           3          1          9   999999-9        2006   
3702930     0           3          1          9   999999-9           0   
3702931     0           3          1          9   999999-9           0   
3702932     0           3          1          9   999999-9           0   
3722739    19           1          1          1   999999-9           0   

        COD_TARIF  REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  \
29749          10    0.00         1       2   850.00        0.00          0   
43148          00    0.00         1       8   150.00        0.00          0   
44185          00    0.00         1       8   127.00        0.00          0   
69066          10    0.00         2       2  8000.00        0.00          0   
96831          10    0.00         2       2  4193.00        0.00          0   
...           ...     ...       ...     ...      ...         ...        ...   
3698859        10    0.00         2       2  5530.00        0.00          0   
3702930        10    0.00         1       1  5509.00        0.00          0   
3702931        10    0.00         1       1  5509.00        0.00          0   
3702932        10    0.00         1       1  5509.00        0.00          0   
3722739        00    0.00         1       8   212.00        0.00          0   

         VAL_RESS  D_RESS     D_AVI     D_LIQ   D_OCORR  CAUSA SEXO  D_NASC  \
29749        0.00       0  20200707  20200708  20090526      9    M       0   
43148        0.00       0  20200707  20200721  20200706      7    0       0   
44185        0.00       0  20201125  20201130  20201109      7    0       0   
69066        0.00       0  20200713  20200714  20081109      9    M       0   
96831        0.00       0  20200717  20200720  20081111      9    M       0   
...           ...     ...       ...       ...       ...    ...  ...     ...   
3698859      0.00       0  20200828  20200923  20090519      9    0       0   
3702930      0.00       0  20201110  20201111  20201023      4    0       0   
3702931      0.00       0  20201110  20201111  20201023      4    0       0   
3702932      0.00       0  20201110  20201111  20201023      4    0       0   
3722739      0.00       0  20200811  20111121  20111026      7    0       0   

         CEP  
29749      0  
43148      0  
44185      0  
69066      0  
96831      0  
...      ...  
3698859    0  
3702930    0  
3702931    0  
3702932    0  
3722739    0  

[637 rows x 22 columns]

Como não será possível recuperar de onde são esses sinistros, vamos retirar essas linhas:

In [31]:
# retirar linhas com 'REGIAO' igual a 0 e com 'CEP' igual a 0
dados = dados.drop(dados[(dados['REGIAO'] == 0) & (dados['CEP'] == 0)].index)

In [32]:
dados.shape

(3747009, 22)

Será que há observações com REGIAO igual a 99 e CEP com valor 0?

In [33]:
# visualizar linhas com REGIAO igual a '99' e CEP igual a 0
dados.query('REGIAO == "99" & CEP == 0')

Empty DataFrame
Columns: [ITEM, MODALIDADE, TIPO_PROD, COBERTURA, COD_MODELO, ANO_MODELO, COD_TARIF, REGIAO, COD_CONT, EVENTO, INDENIZ, VAL_SALVAD, D_SALVADO, VAL_RESS, D_RESS, D_AVI, D_LIQ, D_OCORR, CAUSA, SEXO, D_NASC, CEP]
Index: []

Não há nenhuma linha desse tipo.

In [34]:
dados.shape

(3747009, 22)

Antes tínhamos 3.747.646 observações, agora ficamos com 3.747.009 (pois tiramos as 637 linhas com CEP igual a 0 e região igual a 0).

### 'COD_MODELO'

Outra variável de código que está como string é COD_MODELO. Vamos observá-la:

In [35]:
dados['COD_MODELO'].unique()

array(['004354-0', '014065-1', '013032-0', ..., '021206-7', '529004-0',
       '508052-5'], dtype=object)

Como podemos checar no arquivo "auto2_vei.csv", os códigos de modelos têm realmente essa estrutura:

XXXXXX-X

Então vamos manter como está, sem transformar.

In [36]:
# ordenar os valores das indenizações em ordem decrescente
dados.sort_values('INDENIZ', ascending=False)

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO  \
2053562     5           3          1          9   513163-4        2011   
969269      0           1          2          1   021387-0        2018   
913558      0           1          1          1   021404-3        2019   
3445221     0           1          2          1   035099-0        2018   
3636736   287           3          1          9   513160-0        2011   
...       ...         ...        ...        ...        ...         ...   
3485757     0           1          2          1   005490-9        2020   
3100967     0           1          2          1   509084-9        2006   
2246549     0           1          2          1   003367-7        2013   
2974145     0           1          2          1   022115-5        2015   
3624494     0           1          1          1   004066-5        1998   

        COD_TARIF  REGIAO  COD_CONT  EVENTO    INDENIZ  VAL_SALVAD  D_SALVADO  \
2053562        50    3.00         2       2 1457500.00        0.00          0   
969269         17   40.00         1       1 1050215.00        0.00          0   
913558         17    7.00         1       1 1010027.00        0.00          0   
3445221        17   11.00         1       1 1000501.00        0.00          0   
3636736        50   11.00         1       3 1000000.00        0.00          0   
...           ...     ...       ...     ...        ...         ...        ...   
3485757        10   30.00         1       1  -31083.00        0.00          0   
3100967        40   21.00         1       1  -32995.00        0.00          0   
2246549        10   13.00         1       1  -34684.00        0.00          0   
2974145        22    3.00         1       1  -43506.00        0.00          0   
3624494        10   12.00         1       1 -176275.00     3100.00   20200815   

         VAL_RESS  D_RESS     D_AVI     D_LIQ   D_OCORR  CAUSA SEXO    D_NASC  \
2053562      0.00       0  20200804         0  20190819      9    F         0   
969269       0.00       0  20200825  20201030  20200823      5    M  19750611   
913558       0.00       0  20200731  20201228  20200731      2    F  19650122   
3445221      0.00       0  20200702  20200828  20200701      5    M  19750526   
3636736      0.00       0  20200923         0  20171223      9    0         0   
...           ...     ...       ...       ...       ...    ...  ...       ...   
3485757      0.00       0  20200608         0  20200607      2    M  19780914   
3100967      0.00       0  20200612  20200707  20200610      5    M  19861108   
2246549      0.00       0  20200630         0  20200629      2    M  19720201   
2974145      0.00       0  20200530         0  20200529      4    F  19830129   
3624494      0.00       0  20200717  20200817  20200717      5    0         0   

              CEP  
2053562  88900970  
969269   77024174  
913558   81200110  
3445221   6706230  
3636736   9820280  
...           ...  
3485757         0  
3100967         0  
2246549         0  
2974145         0  
3624494  13091700  

[3747009 rows x 22 columns]

Na página 87 do [manual da SUSEP](https://www.gov.br/susep/pt-br/servicos/mercado/enviar-dados/arquivos/manual_de_orientacao_para_envio_de_dados_out2023.pdf) é possível ter acesso às tabelas com as informações referentes a cada variável.

Variável COD_MODELO está descrita no arquivo 'auto2_vei' (presente na página principal Autoseg ao entrar nos dados do segundo semestre de 2020). Por exemplo, a maior indenização foi:

* Modelo 513163-4: SCANIA - G-420 A 6x4 2p (diesel)
* ANO: 2011
* Indenização: R$ 1.457.500,00
* COD_TARIF: 50 (Rebocadores Nacionais)  
* Evento: 2 (Responsabilidade Civil Facultativa de Veículos - Danos
Materiais)
* CAUSA (Causa geradora do sinistro): 9 (outros)


### 'SEXO'

Há ocorrências "M", "F" e "0" (sem informação).

In [37]:
dados['SEXO'].value_counts()

SEXO
M    1760237
F    1468714
0     518058
Name: count, dtype: int64

In [38]:
# resumo estatístico das linhas com SEXO == 0
dados.query('SEXO == "0"').describe()

ITEM  MODALIDADE  TIPO_PROD  COBERTURA  ANO_MODELO    REGIAO  \
count 518058.00   518058.00  518058.00  518058.00   518058.00 518058.00   
mean     421.68        1.16       1.01       1.37     2003.76     14.40   
std     3617.97        0.53       0.08       1.63      147.02      8.96   
min        0.00        1.00       1.00       1.00        0.00      0.00   
25%        0.00        1.00       1.00       1.00     2012.00      9.00   
50%        1.00        1.00       1.00       1.00     2015.00     13.00   
75%       27.00        1.00       1.00       1.00     2019.00     18.00   
max   180494.00        3.00       2.00       9.00     2022.00     99.00   

       COD_CONT    EVENTO    INDENIZ  VAL_SALVAD   D_SALVADO  VAL_RESS  \
count 518058.00 518058.00  518058.00   518058.00   518058.00 518058.00   
mean       1.02      2.73    4819.00      350.35   471377.78      9.24   
std        0.13      2.87   15880.70     5279.64  3049942.58    725.06   
min        1.00      0.00 -176275.00  -414000.00        0.00 -61585.00   
25%        1.00      1.00     286.00        0.00        0.00      0.00   
50%        1.00      1.00     820.00        0.00        0.00      0.00   
75%        1.00      2.00    3636.00        0.00        0.00      0.00   
max        2.00      8.00 1000000.00   340000.00 20220201.00 200000.00   

           D_RESS       D_AVI       D_LIQ     D_OCORR     CAUSA      D_NASC  \
count   518058.00   518058.00   518058.00   518058.00 518058.00   518058.00   
mean     44504.71 20201655.11 18204059.37 20200892.97      6.52   386965.19   
std     947270.39    56196.94  6033884.36     1593.20      2.43  2738158.42   
min          0.00        0.00        0.00 19970330.00      1.00        0.00   
25%          0.00 20200825.00 20200910.00 20200812.00      4.00        0.00   
50%          0.00 20201015.00 20201116.00 20200928.00      7.00        0.00   
75%          0.00 20201202.00 20210111.00 20201115.00      9.00        0.00   
max   20220303.00 20210228.00 20220317.00 20210224.00      9.00 29221202.00   

              CEP  
count   518058.00  
mean  36946397.04  
std   33601355.32  
min          0.00  
25%    6850720.00  
50%   29060000.00  
75%   74675000.00  
max   99999999.00

É uma decisão do pesquisador manter ou não essas linhas. Como são muitas (518.058), vamos mantê-las por enquanto.

Variável 'D_NASC': não há a informação sobre a idade da pessoa, mas há a data de nascimento, no formato AAAAMMDD.

### Criar nova variável: 'DATA'

In [39]:
dados.head()

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO COD_TARIF  \
0     0           1          1          1   004354-0        2009        10   
1     0           1          1          1   004354-0        2009        10   
2     0           1          1          1   004354-0        2009        10   
3     0           1          1          1   004354-0        2009        10   
4     0           1          1          1   004354-0        2009        10   

   REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  VAL_RESS  D_RESS  \
0   11.00         1       8    63.00        0.00          0      0.00       0   
1   11.00         1       8    63.00        0.00          0      0.00       0   
2   11.00         1       8    80.00        0.00          0      0.00       0   
3   11.00         1       8   100.00        0.00          0      0.00       0   
4   11.00         1       8   127.00        0.00          0      0.00       0   

      D_AVI     D_LIQ   D_OCORR  CAUSA SEXO    D_NASC      CEP  
0  20201016  20201026  20201015      7    M  19740216  9371420  
1  20201017  20201026  20201016      7    M  19740216  9371420  
2  20201017  20201105  20201016      7    M  19740216  9030620  
3  20200812  20200819  20200811      7    M  19740216  9371520  
4  20200812  20200819  20200811      7    M  19740216        0

Vamos criar uma nova coluna chamada 'DATA' para inserir a data de nascimento no formato de data.

In [40]:
dados['DATA'] = pd.to_datetime(dados['D_NASC'], format='%Y%m%d', errors='coerce')

In [41]:
# visualizar os dados
dados.head()

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO COD_TARIF  \
0     0           1          1          1   004354-0        2009        10   
1     0           1          1          1   004354-0        2009        10   
2     0           1          1          1   004354-0        2009        10   
3     0           1          1          1   004354-0        2009        10   
4     0           1          1          1   004354-0        2009        10   

   REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  VAL_RESS  D_RESS  \
0   11.00         1       8    63.00        0.00          0      0.00       0   
1   11.00         1       8    63.00        0.00          0      0.00       0   
2   11.00         1       8    80.00        0.00          0      0.00       0   
3   11.00         1       8   100.00        0.00          0      0.00       0   
4   11.00         1       8   127.00        0.00          0      0.00       0   

      D_AVI     D_LIQ   D_OCORR  CAUSA SEXO    D_NASC      CEP       DATA  
0  20201016  20201026  20201015      7    M  19740216  9371420 1974-02-16  
1  20201017  20201026  20201016      7    M  19740216  9371420 1974-02-16  
2  20201017  20201105  20201016      7    M  19740216  9030620 1974-02-16  
3  20200812  20200819  20200811      7    M  19740216  9371520 1974-02-16  
4  20200812  20200819  20200811      7    M  19740216        0 1974-02-16

In [42]:
# há dados faltantes?
dados.isna().sum()

ITEM               0
MODALIDADE         0
TIPO_PROD          0
COBERTURA          0
COD_MODELO         0
ANO_MODELO         0
COD_TARIF          0
REGIAO             0
COD_CONT           0
EVENTO             0
INDENIZ            0
VAL_SALVAD         0
D_SALVADO          0
VAL_RESS           0
D_RESS             0
D_AVI              0
D_LIQ              0
D_OCORR            0
CAUSA              0
SEXO               0
D_NASC             0
CEP                0
DATA          556912
dtype: int64

As 556.912 linhas com dados faltantes são aquelas linhas onde havia 0 no lugar da data de nascimento.

### Nova variável 'IDADE'

Vamos criar uma coluna nova 'IDADE' para inserir a idade de cada pessoa baseando na data de nascimento (variável 'DATA' já criada):

In [43]:
now = pd.to_datetime('now', utc=True)
dados['IDADE'] = (now.year - dados['DATA'].dt.year) - ((now.month - dados['DATA'].dt.month) < 0)

In [44]:
# visualizar
dados.head()

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO COD_TARIF  \
0     0           1          1          1   004354-0        2009        10   
1     0           1          1          1   004354-0        2009        10   
2     0           1          1          1   004354-0        2009        10   
3     0           1          1          1   004354-0        2009        10   
4     0           1          1          1   004354-0        2009        10   

   REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  VAL_RESS  D_RESS  \
0   11.00         1       8    63.00        0.00          0      0.00       0   
1   11.00         1       8    63.00        0.00          0      0.00       0   
2   11.00         1       8    80.00        0.00          0      0.00       0   
3   11.00         1       8   100.00        0.00          0      0.00       0   
4   11.00         1       8   127.00        0.00          0      0.00       0   

      D_AVI     D_LIQ   D_OCORR  CAUSA SEXO    D_NASC      CEP       DATA  \
0  20201016  20201026  20201015      7    M  19740216  9371420 1974-02-16   
1  20201017  20201026  20201016      7    M  19740216  9371420 1974-02-16   
2  20201017  20201105  20201016      7    M  19740216  9030620 1974-02-16   
3  20200812  20200819  20200811      7    M  19740216  9371520 1974-02-16   
4  20200812  20200819  20200811      7    M  19740216        0 1974-02-16   

   IDADE  
0  51.00  
1  51.00  
2  51.00  
3  51.00  
4  51.00

In [45]:
dados['IDADE'].unique()

array([ 51.,  34.,  61.,  54.,  nan,  46.,  67.,  42.,  74.,  59.,  85.,
        48.,  47.,  41.,  43.,  57.,  64.,  76.,  40.,  36.,  35.,  32.,
        56.,  45.,  38.,  50.,  70.,  33.,  44.,  55.,  68.,  73.,  52.,
        75.,  60.,  66.,  63.,  39.,  58.,  69.,  26.,  37.,  65.,  72.,
        53.,  79.,  49.,  84.,  62.,  31.,  77.,  29.,  81.,  71.,  24.,
        80.,  30.,  78.,  87.,  90.,  83.,  27.,  28.,  91.,  25.,  88.,
        86.,  82.,  97., 103.,  89.,  94.,  92.,  93., 100.,  96.,   4.,
        98.,  23., 102.,  95.,  14., 101.,   7., 124.,   5., 107.,  99.,
        11.,   3.,   6., 105.,  20.,  22.,  21.,  16., 104.,  17.,  18.,
        13.,  19.,   9., 106., 125., 139., 114., -63., 141.,  12., 137.,
       136., 110.,   0.,   2.,   8.,  15., 109.,   1., 343., 146., 227.,
       232., 108., 140., 129., 164., 131., 252., 334., 156., 132.,  10.,
       193., 150.])

Há dados faltantes (nan) e alguns valores que não fazem sentido (valores iguais a 0, negativos ou muito altos).

In [46]:
dados['IDADE'].value_counts()

IDADE
42.00     96359
43.00     92472
44.00     91277
45.00     88331
46.00     84678
          ...  
131.00        1
156.00        1
334.00        1
132.00        1
193.00        1
Name: count, Length: 133, dtype: int64

In [47]:
# apenas visualizar as linhas que contêm nan no lugar de IDADE
# que são as linhas em que D_NASC é 0
dados.query('IDADE.isna()')

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO  \
9           0           1          1          1   015088-6        2019   
18          1           1          1          1   515077-9        2001   
19          1           1          1          1   515077-9        2001   
20          1           1          1          1   515077-9        2001   
21          1           1          1          1   515077-9        2001   
...       ...         ...        ...        ...        ...         ...   
3747628     1           1          1          1   014087-2        2017   
3747629     1           1          1          1   014087-2        2017   
3747630     1           1          1          1   014087-2        2017   
3747641     1           1          1          1   001414-1        2015   
3747642     1           1          1          1   001414-1        2015   

        COD_TARIF  REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  \
9              10   11.00         1       1   536.00        0.00          0   
18             42   13.00         1       1   400.00        0.00          0   
19             42   13.00         1       1    19.00        0.00          0   
20             42   13.00         1       1   320.00        0.00          0   
21             42   13.00         1       1   100.00        0.00          0   
...           ...     ...       ...     ...      ...         ...        ...   
3747628        10   14.00         1       1   703.00        0.00          0   
3747629        10   14.00         1       1  1362.00        0.00          0   
3747630        10   14.00         1       1    90.00        0.00          0   
3747641       14A   36.00         1       2  3090.00        0.00          0   
3747642       14A   36.00         1       2  2300.00        0.00          0   

         VAL_RESS  D_RESS     D_AVI     D_LIQ   D_OCORR  CAUSA SEXO  D_NASC  \
9            0.00       0  20200818  20200917  20200817      9    0       0   
18           0.00       0  20200811  20201008  20200811      9    0       0   
19           0.00       0  20200811  20201105  20200811      9    0       0   
20           0.00       0  20200811  20201008  20200811      9    0       0   
21           0.00       0  20200811  20201008  20200811      9    0       0   
...           ...     ...       ...       ...       ...    ...  ...     ...   
3747628      0.00       0  20200707  20200824  20200622      9    0       0   
3747629      0.00       0  20200623  20200811  20200622      4    0       0   
3747630      0.00       0  20200622  20200804  20200622      9    0       0   
3747641      0.00       0  20200619  20201030  20200413      4    0       0   
3747642      0.00       0  20200619  20200817  20200513      4    0       0   

              CEP DATA  IDADE  
9         7132280  NaT    NaN  
18       17370000  NaT    NaN  
19       17370000  NaT    NaN  
20       17370000  NaT    NaN  
21       17370000  NaT    NaN  
...           ...  ...    ...  
3747628  38420000  NaT    NaN  
3747629  38420000  NaT    NaN  
3747630  38420000  NaT    NaN  
3747641  78705000  NaT    NaN  
3747642  78705000  NaT    NaN  

[556912 rows x 24 columns]

In [48]:
# apenas visualizar as linhas com idade <= 0
dados.query('IDADE <= 0')

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO  \
576722    27           1          1          1   004397-4        2020   
693725     0           1          1          1   811147-2        2019   

       COD_TARIF  REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  \
576722        22   36.00         1       1  8797.00        0.00          0   
693725        30   27.00         1       2  1530.00        0.00          0   

        VAL_RESS  D_RESS     D_AVI     D_LIQ   D_OCORR  CAUSA SEXO    D_NASC  \
576722      0.00       0  20201113  20210118  20201029      4    0  20880122   
693725      0.00       0  20200723  20200817  20200720      4    F  20240924   

             CEP       DATA  IDADE  
576722  78253000 2088-01-22 -63.00  
693725  60130235 2024-09-24   0.00

Poderíamos fazer o mesmo para visualizar as idades maiores do que um determinado valor.

Há problemas com essas idades, mas não vamos alterá-las ainda.

## Visão geral

In [49]:
dados.head()

ITEM  MODALIDADE  TIPO_PROD  COBERTURA COD_MODELO  ANO_MODELO COD_TARIF  \
0     0           1          1          1   004354-0        2009        10   
1     0           1          1          1   004354-0        2009        10   
2     0           1          1          1   004354-0        2009        10   
3     0           1          1          1   004354-0        2009        10   
4     0           1          1          1   004354-0        2009        10   

   REGIAO  COD_CONT  EVENTO  INDENIZ  VAL_SALVAD  D_SALVADO  VAL_RESS  D_RESS  \
0   11.00         1       8    63.00        0.00          0      0.00       0   
1   11.00         1       8    63.00        0.00          0      0.00       0   
2   11.00         1       8    80.00        0.00          0      0.00       0   
3   11.00         1       8   100.00        0.00          0      0.00       0   
4   11.00         1       8   127.00        0.00          0      0.00       0   

      D_AVI     D_LIQ   D_OCORR  CAUSA SEXO    D_NASC      CEP       DATA  \
0  20201016  20201026  20201015      7    M  19740216  9371420 1974-02-16   
1  20201017  20201026  20201016      7    M  19740216  9371420 1974-02-16   
2  20201017  20201105  20201016      7    M  19740216  9030620 1974-02-16   
3  20200812  20200819  20200811      7    M  19740216  9371520 1974-02-16   
4  20200812  20200819  20200811      7    M  19740216        0 1974-02-16   

   IDADE  
0  51.00  
1  51.00  
2  51.00  
3  51.00  
4  51.00

In [50]:
dados.shape

(3747009, 24)